emmet-core/emmet/core/molecules/redox.py

```diff
@@ -0,0 +1,248 @@
+import logging
+from collections import defaultdict
+from typing import Dict, List, Any, Tuple, Union
+import copy
+from collections import defaultdict
+
+from typing_extensions import Literal
+
+import numpy as np
+from pydantic import Field
+import networkx as nx
+
+from pymatgen.core.periodic_table import Specie, Element
+from pymatgen.core.structure import Molecule
+from pymatgen.analysis.graphs import MoleculeGraph
+from pymatgen.analysis.local_env import OpenBabelNN, metal_edge_extender
+from pymatgen.analysis.molecule_matcher import MoleculeMatcher
+
+from emmet.core import SETTINGS
+from emmet.core.mpid import MPID
+from emmet.core.qchem.task import TaskDocument
+from emmet.core.qchem.calc_types import TaskType
+from emmet.core.qchem.molecule import evaluate_lot
+from emmet.core.material import PropertyOrigin
+from emmet.core.molecules.molecule_property import PropertyDoc
+from emmet.core.molecules.bonds import metals
+from emmet.core.molecules.thermo import get_free_energy
+from emmet.core.mpid import MPID
+
+
+
+reference_potentials = {"H": 4.44,
+                        "Li": 1.40,
+                        "Mg": 2.06,
+                        "Ca": 1.60}
+
+class RedoxDoc(PropertyDoc):
+    """
+    Molecular properties related to reduction and oxidation, including
+    vertical ionization energies and electron affinities, as well as reduction
+    and oxidation potentials
+    """
+
+    property_name = "redox"
+
+    electron_affinity: float = Field(description="Vertical electron affinity in eV")
+
+    ea_id: MPID = Field(description="MPID for electron affinity")
+
+    ionization_energy: float = Field(description="Vertical ionization energy in eV")
+
+    ie_id: MPID = Field(description="MPID for ionization energy")
+
+    reduction_free_energy: float = Field(None, description="Adiabatic free energy of reduction")
+
+    red_id: MPID = Field(None, description="MPID for adiabatic reduction")
+
+    oxidation_free_energy: float = Field(None, description="Adiabatic free energy of oxidation")
+
+    ox_id: MPID = Field(None, description="MPID for adiabatic oxidation")
+
+    reduction_potentials: Dict[str, float] = Field(None, description="Reduction potentials with various reference electrodes")
+
+    oxidation_potentials: Dict[str, float] = Field(None, description="Oxidation potentials with various reference electrodes")
+
+    @classmethod
+    def from_entries(
+        cls,
+        entries: List[Dict[str, Any]],
+        **kwargs
+    ) -> List["RedoxDoc"]:
+        """
+        Construct documents describing molecular redox properties from task documents.
+        Note that multiple documents may be made by this procedure.
```

Comment:
It's correct. We don't have a ComputedEntry equivalent for molecules, so I defined a schema for dictionaries.

Reply:
Thanks for clarifying. What I meant was that the method is called `from_entries` but the docstring says it constructs documents "from task documents." If it really builds from task documents, shouldn't the type hint be `entries: [TaskDocument]` rather than `entries: List[Dict[str, Any]]`?